zkevm-circuits/src/keccak_circuit/test.rs

```diff
@@ -1,12 +1,22 @@
-use crate::keccak_circuit::KeccakCircuit;
-use eth_types::Field;
+use super::*;
+use crate::evm_circuit::util::rlc;
+use halo2_proofs::dev::CellValue;
+use halo2_proofs::plonk::Assignment;
 use halo2_proofs::{dev::MockProver, halo2curves::bn256::Fr};
+use itertools::izip;
 use log::error;
 
-fn verify<F: Field>(k: u32, inputs: Vec<Vec<u8>>, success: bool) {
-    let circuit = KeccakCircuit::new(Some(2usize.pow(k)), inputs);
-
+fn verify<F: Field>(k: u32, inputs: Vec<Vec<u8>>, digests: Vec<String>, success: bool) {
+    let circuit = KeccakCircuit::new(Some(2usize.pow(k)), inputs.clone());
     let prover = MockProver::<F>::run(k, &circuit, vec![]).unwrap();
+    let (config, challenges) = KeccakCircuit::configure(&mut ConstraintSystem::<F>::default());
+    let input_challenge = prover
+        .get_challenge(challenges.keccak_input())
+        .inner
+        .unwrap();
+    let hash_challenge = prover.get_challenge(challenges.evm_word()).inner.unwrap();
+
```

Comment:
extracting challenges from **unassigned** circuit is anti-pattern.  Although currently it can pass **mock** proving, it is against what challenges should behave.   Currently MockProver uses pre-defined challenges which are not related to advice at all. We'd better insert circuit synthesis step between line12 and line13

Reply:
These are the challenges given by the prover to the circuit and to the test. Synthesis happens in `MockProver::run`. The MockProver can use real challenges too as long as it’s get_challenge method is consistent.

Reply:
i guess the codes above probably will not work any more after we sync https://github.com/privacy-scaling-explorations/halo2/pull/129/files in the future?

Reply:
This should work (in principle, I didn’t try this branch yet).